powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/ActivationServiceBehavior.java

```diff
@@ -1105,63 +1142,229 @@ public CreateActivationResponse createActivation(
+                recoveryCodeRepository.save(recoveryCodeEntity);
             }
+        }
 
-        } else {
+        CommitActivationResponse response = new CommitActivationResponse();
+        response.setActivationId(activationId);
+        response.setActivated(true);
+        return response;
+    }
+
+    /**
+     * Update activation OTP for given activation ID.
+     *
+     * @param activationId Activation ID.
+     * @param externalUserId User ID of user who committed the activation. Use null value if activation owner caused the change.
+     * @param activationOtp Activation OTP.
+     * @return Response with activation UTP update result.
+     * @throws GenericServiceException In case invalid data is provided or activation is not found, in invalid state or already expired.
+     */
+    public UpdateActivationOtpResponse updateActivationOtp(String activationId, String externalUserId, String activationOtp) throws GenericServiceException {
+
+        // Validate provided OTP
+        if (activationOtp == null || activationOtp.isEmpty()) {
+            logger.warn("Activation OTP not specified in update");
+            // Rollback is not required, error occurs before writing to database
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
+        }
+
+        // Get the repository
+        final ActivationRepository activationRepository = repositoryCatalogue.getActivationRepository();
+
+        // Find activation
+        ActivationRecordEntity activation = activationRepository.findActivationWithLock(activationId);
+        if (activation == null) {
             // Activation does not exist
             logger.info("Activation does not exist, activation ID: {}", activationId);
-            // Rollback is not required, database is not used for writing
             throw localizationProvider.buildExceptionForCode(ServiceError.ACTIVATION_NOT_FOUND);
         }
+
+        // Get current timestamp
+        final Date timestamp = new Date();
+
+        // Check already deactivated activation
+        deactivatePendingActivation(timestamp, activation, true);
+
+        // Check activation state
+        if (activation.getActivationStatus() != ActivationStatus.OTP_USED) {
+            logger.info("Activation is not in OTP_USED state during commit, activation ID: {}", activationId);
+            // Rollback is not required, database is not used for writing
+            throw localizationProvider.buildExceptionForCode(ServiceError.ACTIVATION_INCORRECT_STATE);
+        }
+
+        // Check OTP validation mode
+         if (activation.getActivationOtpValidation() == io.getlime.security.powerauth.app.server.database.model.ActivationOtpValidation.ON_KEY_EXCHANGE) {
+            logger.info("Activation OTP update is not allowed for ON_KEY_EXCHANGE mode. Activation ID: {}", activationId);
+            // Rollback is not required, database is not used for writing
```

Comment:
The comment `// Rollback is not required, error occurs before writing to database` is more relevant, because the method writes to DB later on in case of no error.